### ISSUE TITLE
OData function no longer working as expected

### ISSUE THREAD

**Issue by gordon-matt** [OData/AspNetCoreOData#444]
I have just updated a project to NET 6 and OData 8.0.6.. I have a function that is no longer working as expected.

**Method Definition**:

```
public virtual async Task<IActionResult> GetComparitiveTable(
    [FromODataUri] string cultureCode,
    ODataQueryOptions<ComparitiveLocalizableString> options)
{
    if (!CheckPermission(ReadPermission))
    {
        return Unauthorized();
    }
    else
    {
        int tenantId = GetTenantId();
        var connection = GetDisposableConnection();

        var query = connection.Query(x => x.TenantId == tenantId && (x.CultureCode == null || x.CultureCode == cultureCode))
		    .ToHashSet()
            .GroupBy(x => x.TextKey)
            .Select(grp => new ComparitiveLocalizableString
            {
                Key = grp.Key,
                InvariantValue = grp.First(x => x.CultureCode == null).TextValue,
                LocalizedValue = grp.FirstOrDefault(x => x.CultureCode == cultureCode) == null
                    ? string.Empty
                    : grp.First(x => x.CultureCode == cultureCode).TextValue
            })
            .AsQueryable();

        var results = options.ApplyTo(query, IgnoreQueryOptions);
        var response = await Task.FromResult((results as IQueryable<ComparitiveLocalizableString>).ToHashSet());
        return Ok(response);
    }
}
```

Note that an entity is queried (`LocalizableString`) and a complex type is returned (`ComparitiveLocalizableString`).

**Definition of** `LocalizableString`:

```
public class LocalizableString : TenantEntity<Guid>
{
	public string CultureCode { get; set; }

	public string TextKey { get; set; }

	public string TextValue { get; set; }
}
```

**Definition of** `ComparitiveLocalizableString`:

```
public class ComparitiveLocalizableString
{
    public string Key { get; set; }

    public string InvariantValue { get; set; }

    public string LocalizedValue { get; set; }
}
```

**OData Function Setup**:

```
var getComparitiveTableFunction = builder.EntityType<LocalizableString>().Collection.Function("GetComparitiveTable");
getComparitiveTableFunction.Parameter<string>("cultureCode");
getComparitiveTableFunction.Returns<IActionResult>();
```

**URL**:

`${this.apiUrl}/Default.GetComparitiveTable(cultureCode='${this.cultureCode}')`

**Error**:

```
Microsoft.OData.ODataException: The key value (Default.GetComparitiveTable(cultureCode='en-US')) from request is not valid. The key value should be format of type 'Edm.Guid'.
```

I'm not trying to query by the entity's key, but by it's `CultureCode`, as I have specified in the OData function definition (and which has been working fine for several versions of **AspNetCoreOData** until now).

**Comment by gordon-matt** [comment]
Note I am seeing more problems with OData functions when updating another project as well. In this case, it's a 404 I get. Here are the details:

**Method Definition**:

```
public async Task<IActionResult> GetFlightRows(
	[FromODataUri] string timeZoneType,
	[FromODataUri] string bookingDateFrom,
	[FromODataUri] string bookingDateTo,
	[FromODataUri] string travelDateFrom,
	[FromODataUri] string travelDateTo,
	[FromODataUri] string currencyCode,
	[FromODataUri] string siteId,
	[FromODataUri] string bookingStatuses,
	[FromODataUri] string paymentStatuses,
	ODataQueryOptions<SupplierProductionReviewReportRow> options)
{
	// etc..
}
```

**OData Function Setup**:

```
var getFlightRows = builder.EntityType<SupplierProductionReviewReportRow>().Collection.Function("GetFlightRows");
getFlightRows.Parameter<string>("timeZoneType");
getFlightRows.Parameter<string>("bookingDateFrom");
getFlightRows.Parameter<string>("bookingDateTo");
getFlightRows.Parameter<string>("travelDateFrom");
getFlightRows.Parameter<string>("travelDateTo");
getFlightRows.Parameter<string>("currencyCode");
getFlightRows.Parameter<string>("siteId");
getFlightRows.Parameter<string>("bookingStatuses");
getFlightRows.Parameter<string>("paymentStatuses");
getFlightRows.Returns<IActionResult>();
```

**URL**:

```
/odata/dashboard/SupplierProductionReviewReportApi/Default.GetFlightRows(timeZoneType='0',bookingDateFrom='2022-01-01',bookingDateTo='2022-01-14',travelDateFrom=null,travelDateTo=null,currencyCode='USD',siteId=null,bookingStatuses='2',paymentStatuses='30')?$orderby=BookingYear,BookingMonth,Country,Destination,BookingId desc&$top=25&$count=true
```

**Comment by julealgon** [comment]
Can you check if your functions show up in the OData route debug view?

On your first example, it is clearly trying to pass everything after `${this.apiUrl}/` as the key (for example `this.apiUrl/1234`) instead of calling the function. This could happen if the function is just not being detected, for example.

**Comment by gordon-matt** [comment]
@julealgon I checked and I see it showing up under the `Non-OData Endpoint Mappings` with a template of `{controller=Home}/{action=Index}/{id?}`. So yeah.. definitely some issue mapping that one and some others as well

**Comment by julealgon** [comment]
@gordon-matt , is your controller inheriting from `ODataController`, or is it still using `ControllerBase`? The underlying base class can affect how attribute/conventional routing works. It's possible that MVC is "winning" vs OData in your case.

Does it work fine if you add the `HttpPost` attribute to it with the full explicit route (effectively, migrating from conventional routing to OData attribute routing)?

**Comment by gordon-matt** [comment]
@julealgon It inherits from `ODataController`.. and as I said, this has been working fine until now. In fact, I've been using this code since 2018.. no change on my part .. and especially no change now, other than upgrading to .NET 6 and OData v8.

**Comment by julealgon** [comment]
@gordon-matt 
1. Are you using conventional MVC routing with a route template? Those defaults suggest you are, just want to make sure.
2. Can you show the controller definition including all it's attributes?
3. Can you show the `UseEndpoints` and `AddOData` sections on your startup?
4. Does any other route show as a proper OData route in the route debug view (assuming you have other OData methods)?
5. Does the problem go away if you change from `IActionResult` to `ActionResult<T>` with a strongly typed `T` and then use just `T` in the EDM?
6. Does the problem go away if you replace:
```csharp
getComparitiveTableFunction.Returns<IActionResult>();
```
with
```csharp
getComparitiveTableFunction.Returns<Task<IActionResult>>();
```
?

**Comment by gordon-matt** [comment]
Okay, to make things easier for you guys to investigate, I created an [example repo](https://github.com/gordon-matt/ODataIssue444). Instructions in the README

**Comment by julealgon** [comment]
> Okay, to make things easier for you guys to investigate, I created an [example repo](https://github.com/gordon-matt/ODataIssue444). Instructions in the README

@gordon-matt The function works normally if I add the following attribute to it:
```csharp
[HttpGet("odata/LocalizableStringApi/Default.GetComparitiveTable(cultureCode={cultureCode})")]
public virtual async Task<IActionResult> GetComparitiveTable(
...
```

![image](https://user-images.githubusercontent.com/461579/149791067-8ac21cc7-1379-483c-9f5d-b8d09b82c319.png)

 It looks like something changed in the convention routing logic that stopped detecting your function implicitly.

I'll see if I can spot what is going on. In the meantime, you might just want to add the explicit template route to unblock yourself.

**Comment by julealgon** [comment]
Found the missing piece. It seems that in OData v8, functions need the `[HttpGet]` attribute to be considered in the convention scan.

If you add this to the function, it starts being detected:
```csharp
[HttpGet]
public virtual async Task<IActionResult> GetComparitiveTable(
....
```

I don't quite understand why the restriction was added, considering we are talking about "convention" and not explicit attribute routing.

Here is the offending check:
https://github.com/OData/AspNetCoreOData/blob/69eec03c7003fe12d92cdc619efdc16781683694/src/Microsoft.AspNetCore.OData/Routing/Conventions/OperationImportRoutingConvention.cs#L90-L106

There is a TODO comment in there directly related to this. @xuzhg you might want to take a look at this one.

**Comment by xuzhg** [comment]
It seems we should remove [HttpGe] and [HttpPost] attributes check in convention routing！

**Comment by gordon-matt** [comment]
Now that I know what the issue is, I don't think it's necessary to remove those attributes, unless you want to keep the same behavior as previous versions.

Apparently some of my old functions had the `[HttpGet]` attribute, but many did not.. and OData didn't care whether it was there or not.. it just worked.

I've now decorated all my functions with `[HttpGet]` for v8 and they work fine again. So this is not exactly a priority issue. If you want to support the old behavior, by all means go ahead.. but it may not be necessary. Perhaps a note in the documentation about this breaking change may suffice.

**Comment by julealgon** [comment]
> It seems we should remove [HttpGe] and [HttpPost] attributes check in convention routing！

This feels like the best approach to me. Even if adding the attribute fixes the problem, there is no particular reason to even introduce the difference to begin with. If the attribute is not needed for "normal" actions (entity sets etc) then it follows that it also shouldn't be needed for functions and actions. This would make the behavior more consistent for the end user.

Additionally, it seems we are not checking the properties of the attribute as well. So someone could add a `[HttpGet("thisIsCompletelyNonsensical")]` and it would be picked up by the convention as a completely different route. That makes no sense and would cause extra confusion.

**Comment by henrikdahl8240** [comment]
@xuzhg Isn't the case, that [HttpGet] and [HttpPost] are needed for the routing, so there is not so much point in removing them from the convention scan. I also think it's just excellent to be explicit in this regard as it's currently for 8.0.7.

**Comment by julealgon** [comment]
> @xuzhg Isn't the case, that [HttpGet] and [HttpPost] are needed for the routing, so there is not so much point in removing them from the convention scan. I also think it's just excellent to be explicit in this regard as it's currently for 8.0.7.

@henrikdahl8240 OData supports 2 ways of identifying routes: using conventions, and using attributes. Conventions historically did not need attributes to work as long as you named the methods as per the expected patterns.

**Comment by henrikdahl8240** [comment]
@julealgon Yes, this is also how I did it earlier, but wasn't it explained in an article at the OData blog, that the decision to make it explicit was driven by the intention, that this is how it is in modern Web Api and OData should not introduce special cases, which you will then need to remember?

**Comment by julealgon** [comment]
> @julealgon Yes, this is also how I did it earlier, but wasn't it explained in an article at the OData blog, that the decision to make it explicit was driven by the intention, that this is how it is in modern Web Api and OData should not introduce special cases, which you will then need to remember?

I don't think so, otherwise the convention approach would've been removed outright. The only change was that in v7-, there were dedicated attributes for OData routes, and in v8 they are now using the same attributes as MVC uses.

You still have the option of not having any attributes if you want to.

I'm not saying necessarily that I think this is good, or that I agree with it, it's just the way it is.
